crates/pico_macros/src/memo.rs

```diff
@@ -0,0 +1,134 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, FnArg, ItemFn, PatType, ReturnType};
+
+pub(crate) fn memo(_args: TokenStream, item: TokenStream) -> TokenStream {
+    let ItemFn {
+        sig,
+        vis,
+        block,
+        attrs,
+    } = parse_macro_input!(item as ItemFn);
+
+    let fn_name = sig.ident.clone().to_string();
+
+    let inputs = sig.inputs.iter().collect::<Vec<_>>();
+
+    if inputs.is_empty() {
+        return syn::Error::new_spanned(&sig, "Function must have at least one argument")
+            .to_compile_error()
+            .into();
+    }
+
+    let db_arg = match &inputs[0] {
+        FnArg::Typed(PatType { pat, .. }) => pat.clone(),
+        _ => {
+            return syn::Error::new_spanned(inputs[0], "Expected a typed argument")
+                .to_compile_error()
+                .into()
+        }
+    };
+
+    let other_args = inputs.iter().skip(1).map(|arg| match arg {
+        FnArg::Typed(PatType { pat, .. }) => pat,
+        _ => unreachable!(),
+    });
+
+    let argument_types = inputs.iter().skip(1).map(|arg| match arg {
+        FnArg::Typed(PatType { ty, .. }) => ty,
+        _ => unreachable!(),
+    });
+
+    let return_type = match &sig.output {
+        ReturnType::Type(_, ty) => ty.clone(),
+        ReturnType::Default => {
+            return syn::Error::new_spanned(
+                &sig.output,
+                "Memoized function must have a return type",
+            )
+            .to_compile_error()
+            .into()
+        }
+    };
+
+    let (return_expr, return_type) = if let syn::Type::Reference(type_ref) = *return_type {
+        let base_type = type_ref.elem;
+        (
+            quote! {
+                #db_arg
+                    .values
+                    .get(&node_id)
+                    .expect("value should exist. This is indicative of a bug in Isograph.")
+                    .as_any()
+                    .downcast_ref::<#base_type>()
+                    .expect("unexpected return type. This is indicative of a bug in Isograph.")
+            },
+            base_type,
+        )
+    } else {
+        (
+            quote! {
+                #db_arg
+                    .values
+                    .get(&node_id)
+                    .expect("value should exist. This is indicative of a bug in Isograph.")
+                    .as_any()
+                    .downcast_ref::<#return_type>()
+                    .expect("unexpected return type. This is indicative of a bug in Isograph.")
+                    .clone()
+            },
+            return_type,
+        )
+    };
+
+    let output = if inputs.len() == 2 {
```

Comment:
do we need to special case this? In the non-length-two case, we wrap the other args in a tuple. Can we do the same here (i.e. a length-one tuple)?

Reply:
I thought what it would be a better solution here and went for a special case. But length-one tuple could simplify code, let me try it and see how it goes

Reply:
It should generate the same code... I think! A length-one tuple is the same at runtime (in terms of memory etc) and just differs at the type level. e.g. it's a newtype